File: src/com/java/test/FactoryStatic.java
package com.java.test;

/**
 * 简单工厂
 * 工厂类负责创建其他类的实例，被创建的实例有共同的父类
 */
public class FactoryStatic {

    public static Cpu provide(int type) {
        if (type == 1) return new CpuAmd();
        if (type == 2) return new CpuIntel();
        throw new RuntimeException("error type");
    }


    public static void main(String[] args) {
        Cpu cpu = FactoryStatic.provide(1);
        cpu.cal();
    }

}


interface Cpu {

    void cal();

}

class CpuAmd implements Cpu {

    @Override
    public void cal() {
        System.out.println("AMD cal...");
    }
}

class CpuIntel implements Cpu {

    @Override
    public void cal() {
        System.out.println("Intel cal...");
    }
}

File: src/com/kotlin/test/FactoryStatic.kt
package com.kotlin.test

/**
 * 静态工厂方法(简单工厂)
 */
fun main(args: Array<String>) {

    FactoryStatic()

}

class FactoryStatic {

    init {
        var printer: Printer? = PrinterFactory.provide("usb")
        printer?.print("data")
    }

    interface Printer {
        fun print(data: String)
    }

    class UsbPrinter : Printer {
        override fun print(data: String) {
            println("UsbPrinter data=$data")
        }
    }

    class BlePrinter : Printer {
        override fun print(data: String) {
            println("BlePrinter data=$data")
        }

    }

    /**
     * 调用者可以免除直接创建对象的职责
     */
    class PrinterFactory {
        companion object {
            fun provide(type: String): Printer? {
                when (type) {
                    "usb" -> return UsbPrinter()
                    "ble" -> return BlePrinter()
                    else -> {
                        throw RuntimeException("类型错误")
                    }
                }
                return null
            }
        }


    }
}
